pkg/kvstore/etcd.go

```diff
@@ -187,20 +187,14 @@ type clientOptions struct {
 func (e *etcdModule) newClient(ctx context.Context, opts *ExtraOptions) (BackendOperations, chan error) {
 	errChan := make(chan error, 10)
 
-	endpointsOpt, endpointsSet := e.opts[EtcdAddrOption]
-	configPathOpt, configSet := e.opts[EtcdOptionConfig]
-
-	rateLimitOpt, rateLimitSet := e.opts[EtcdRateLimitOption]
-
 	clientOptions := clientOptions{
 		KeepAliveHeartbeat: 15 * time.Second,
 		KeepAliveTimeout:   25 * time.Second,
 		RateLimit:          defaults.KVstoreQPS,
 	}
 
-	if rateLimitSet {
-		// error is discarded here because this option has validation
-		clientOptions.RateLimit, _ = strconv.Atoi(rateLimitOpt.value)
+	if o, ok := e.opts[EtcdRateLimitOption]; ok && o.value != "" {
+		clientOptions.RateLimit, _ = strconv.Atoi(o.value)
 	}
```

Comment:
Wait a second, why is `rateLimitOpt` empty here? Wouldn't `rateLimitSet` be `false`? If it is `true` how is `rateLimitOpt` set to `""`?

Reply:
`rateLimitSet` (or `ok` in this case) will be `true` whenever the key `EtcdRateLimitOption` is present, regardless of the value stored in the map: https://play.golang.org/p/-GOb-BcWXMe

Reply:
Hi @aanm , according to my understanding, `e.opts[EtcdRateLimitOption]` is initialized here: https://github.com/cilium/cilium/blob/master/pkg/kvstore/etcd.go#L135, with the `value` field unspecified. So when it comes to https://github.com/cilium/cilium/blob/master/pkg/kvstore/etcd.go#L193, 

```go
    rateLimitOpt, rateLimitSet := e.opts[EtcdRateLimitOption]
```

`rateLimitSet` will be `true` but `rateLimitOpt.value` will be `""`.

Reply:
@tklauser my question was more like how did the map end up having an empty value set to it. @ArthurChiao just explain it. Thanks!

Reply:
This also means that `endpointSet` and `configSet` have the same issue?

Reply:
> This also means that endpointSet and configSet have the same issue?

Yes, I think they are. But, for these two, the problem is hidden as they always use values from CLI inputs:

1.  kvstore.Setup() will pass the CLI parameters to kvstore package
2. `etcd.config` from CLI: defaults to `/var/lib/etcd-config/etcd.config`
2. `etcd.endpoints`: parsed from the config file `etcd.config` (e.g. `/var/lib/etcd-config/etcd.config`)

So the final values are correct (which hide the buggy code).

@aanm

Reply:
Should I fix them in this patch as well?

Reply:
That would be great! It can be done as a separate commit.